# Lab 5: Neural Networks

GSB 545

Week 6

Stephanie Liu

**Predicting Income**

In this lab we'll be using a dataset from kaggle yet again...it's just so fun and rich! We're using the following income dataset where we want to use the other features to predict whether someone is making over $50,000 per year or not.


**Primary Goals:**

Predict income. We want to predict whether an individual makes more than $50,000. We'll use the variables in the dataset to help predict (classify) income.

We will use neural networks to answer the question above, and we'll also explore at least two other models in order to answer the above questions as best we can.


## Data

Let's read in the data and take a closer look at the income dataset and the variables that are contained within it.

In [ ]:
import pandas as pd
income_data = pd.read_csv("/content/income_evaluation.csv")
income_data

age          workclass   fnlwgt    education   education-num  \
0       39          State-gov    77516    Bachelors              13   
1       50   Self-emp-not-inc    83311    Bachelors              13   
2       38            Private   215646      HS-grad               9   
3       53            Private   234721         11th               7   
4       28            Private   338409    Bachelors              13   
...    ...                ...      ...          ...             ...   
32556   27            Private   257302   Assoc-acdm              12   
32557   40            Private   154374      HS-grad               9   
32558   58            Private   151910      HS-grad               9   
32559   22            Private   201490      HS-grad               9   
32560   52       Self-emp-inc   287927      HS-grad               9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex   capital-gain   capital-loss   hours-per-week  native-country  \
0         Male           2174              0               40   United-States   
1         Male              0              0               13   United-States   
2         Male              0              0               40   United-States   
3         Male              0              0               40   United-States   
4       Female              0              0               40            Cuba   
...        ...            ...            ...              ...             ...   
32556   Female              0              0               38   United-States   
32557     Male              0              0               40   United-States   
32558   Female              0              0               40   United-States   
32559     Male              0              0               20   United-States   
32560   Female          15024              0               40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

### Exploration

Now that we've read in the data, we'll take a closer look and try to understand it a little better.

In [ ]:
income_data.dtypes

age                 int64
 workclass         object
 fnlwgt             int64
 education         object
 education-num      int64
 marital-status    object
 occupation        object
 relationship      object
 race              object
 sex               object
 capital-gain       int64
 capital-loss       int64
 hours-per-week     int64
 native-country    object
 income            object
dtype: object

In [ ]:
# rename all variables that have a space in front
income_data.rename(columns={' workclass': 'workclass', 
                            ' fnlwgt': 'fnlwgt', 
                            ' education' : 'education', 
                            ' education-num' : 'education-num', 
                            ' marital-status' : 'marital-status', 
                            ' occupation' : 'occupation', 
                            ' relationship' : 'relationship', 
                            ' race' : 'race', 
                            ' sex' : 'sex',
                            ' capital-gain' : 'capital-gain', 
                            ' capital-loss' : 'capital-loss', 
                            ' hours-per-week' : 'hours-per-week', 
                            ' native-country' : 'native-country', 
                            ' income' : 'income'}, inplace=True)

We see that there are 15 variables in the dataset, and 32,561 rows. We want to predict the *income* variable, which is either "<=50K" or ">50K". 

In [ ]:
# summary statistics for numeric variables
income_data.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

There are 6 numeric variables, and we can see their summary statistics above. Note that education-num is a numeric level assigned to the categorical education level. Since they are measuring the same thing, we may leave one of them out in our modeling efforts.

In [ ]:
# workclass category breakdown
print(income_data['workclass'].value_counts())
print('--')

# education category breakdown
print(income_data['education'].value_counts())
print('--')

# marital status category breakdown
print(income_data['marital-status'].value_counts())
print('--')

# occupation category breakdown
print(income_data['occupation'].value_counts())
print('--')

# relationship category breakdown
print(income_data['relationship'].value_counts())
print('--')

# race category breakdown
print(income_data['race'].value_counts())
print('--')

# sex category breakdown
print(income_data['sex'].value_counts())
print('--')

# native-country category breakdown
print(income_data['native-country'].value_counts())
print('--')

# income category breakdown
print(income_data['income'].value_counts())
print('--')

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64
--
 HS-grad         10501
 Some-college     7291
 Bachelors        5355
 Masters          1723
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           646
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           333
 1st-4th           168
 Preschool          51
Name: education, dtype: int64
--
 Married-civ-spouse       14976
 Never-married            10683
 Divorced                  4443
 Separated                 1025
 Widowed                    993
 Married-spouse-absent      418
 Married-AF-spouse           23
Name: marital-status, dtype: int64
--
 Prof-specialty       4140
 Craft-repair         4099
 Exec-m

In [ ]:
income_data.describe(include='object')

workclass education       marital-status       occupation relationship  \
count      32561     32561                32561            32561        32561   
unique         9        16                    7               15            6   
top      Private   HS-grad   Married-civ-spouse   Prof-specialty      Husband   
freq       22696     10501                14976             4140        13193   

          race    sex  native-country  income  
count    32561  32561           32561   32561  
unique       5      2              42       2  
top      White   Male   United-States   <=50K  
freq     27816  21790           29170   24720

There are 9 categorical variables (including the target variable *income*). 
- workclass has 9 different classes.
- education has 16 different classes.
- marital-status has 7 different classes.
- occupation has 25 different classes.
- relationship has 6 different classes.
- race has 5 different classes.
- sex has 2 different classes.
- native-country has 42 different classes.
- income has 2 different classes (above or below 50k).

In [ ]:
# check for na values
income_data.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

It looks like there are no NA values in the dataset, so we do not need to worry about removing any rows or adjusting for this.

### Data Cleaning and Prepping

Recall that our goal is to predict whether or not a person makes over $50,000. We'll make a few adjustments to the data in order to prepare it for modeling.

Just to make the income categories a little clearer, we'll replace "<=50K" with "less than 50K" and ">50K" with "greater than 50K".

In [ ]:
income_data.loc[(income_data.income == ' <=50K'),'income']='less than 50K'
income_data.loc[(income_data.income == ' >50K'),'income']='greater than 50K'

# income category breakdown
print(income_data['income'].value_counts())

less than 50K       24720
greater than 50K     7841
Name: income, dtype: int64


Let's also change the "?" classes in workclass, occupation and native-country to "unknown".

In [ ]:
income_data.loc[(income_data.workclass == ' ?'),'workclass']=' Unknown'
print(income_data['workclass'].value_counts())
print('---')

income_data.loc[(income_data.occupation == ' ?'),'occupation']=' Unknown'
print(income_data['occupation'].value_counts())
print('---')

income_data.loc[(income_data['native-country'] == ' ?'),'native-country']=' Unknown'
print(income_data['native-country'].value_counts())
print('---')

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 Unknown              1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64
---
 Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 Unknown              1843
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
Name: occupation, dtype: int64
---
 United-States                 29170
 Mexico                          643
 Unknown                         583
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador            

For the Neural Networks Modeling, we need all numeric inputs. Therefore, we should create dummy variables for the categorical variables. Since we will use the numeric coding for education, we will not create dummies for the categorical education variable.

In [ ]:
# create dummy variables for the categorical variables
income_data_clean = income_data.merge(pd.get_dummies(income_data[['workclass', 
                                                                  'marital-status',
                                                                  'occupation', 
                                                                  'relationship',
                                                                  'race',
                                                                  'sex', 
                                                                  'native-country']]), 
                                      left_index = True, right_index = True)

# shift column 'income' to first position
first_column = income_data_clean.pop('income') 
# insert column using insert(position,column_name,first_column) function
income_data_clean.insert(0, 'income', first_column)

income_data_clean.head()

income  age          workclass  fnlwgt   education  education-num  \
0  less than 50K   39          State-gov   77516   Bachelors             13   
1  less than 50K   50   Self-emp-not-inc   83311   Bachelors             13   
2  less than 50K   38            Private  215646     HS-grad              9   
3  less than 50K   53            Private  234721        11th              7   
4  less than 50K   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race  ...  \
0        Never-married        Adm-clerical   Not-in-family   White  ...   
1   Married-civ-spouse     Exec-managerial         Husband   White  ...   
2             Divorced   Handlers-cleaners   Not-in-family   White  ...   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black  ...   
4   Married-civ-spouse      Prof-specialty            Wife   Black  ...   

  native-country_ Puerto-Rico  native-country_ Scotland  \
0                           0                         0   
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0                         0   

   native-country_ South  native-country_ Taiwan native-country_ Thailand  \
0                      0                       0                        0   
1                      0                       0                        0   
2                      0                       0                        0   
3                      0                       0                        0   
4                      0                       0                        0   

   native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                0                              1   
1                                0                              1   
2                                0                              1   
3                                0                              1   
4                                0                              0   

   native-country_ Unknown  native-country_ Vietnam  \
0                        0                        0   
1                        0                        0   
2                        0                        0   
3                        0                        0   
4                        0                        0   

   native-country_ Yugoslavia  
0                           0  
1                           0  
2                           0  
3                           0  
4                           0  

[5 rows x 101 columns]

We see that there are now 101 columns (variables) after each categorical variable has been changed into dummy variables.

Now, we would like to remove the original categorical variables, since we will not be including them in our models.

In [ ]:
income_data_clean = income_data_clean.drop(['workclass', 
                                            'education',
                                            'marital-status',
                                            'occupation', 
                                            'relationship',
                                            'race',
                                            'sex', 
                                            'native-country'], axis = 1)

income_data_clean.head()

income  age  fnlwgt  education-num  capital-gain  capital-loss  \
0  less than 50K   39   77516             13          2174             0   
1  less than 50K   50   83311             13             0             0   
2  less than 50K   38  215646              9             0             0   
3  less than 50K   53  234721              7             0             0   
4  less than 50K   28  338409             13             0             0   

   hours-per-week  workclass_ Federal-gov  workclass_ Local-gov  \
0              40                       0                     0   
1              13                       0                     0   
2              40                       0                     0   
3              40                       0                     0   
4              40                       0                     0   

   workclass_ Never-worked  ...  native-country_ Puerto-Rico  \
0                        0  ...                            0   
1                        0  ...                            0   
2                        0  ...                            0   
3                        0  ...                            0   
4                        0  ...                            0   

   native-country_ Scotland  native-country_ South  native-country_ Taiwan  \
0                         0                      0                       0   
1                         0                      0                       0   
2                         0                      0                       0   
3                         0                      0                       0   
4                         0                      0                       0   

   native-country_ Thailand  native-country_ Trinadad&Tobago  \
0                         0                                0   
1                         0                                0   
2                         0                                0   
3                         0                                0   
4                         0                                0   

   native-country_ United-States  native-country_ Unknown  \
0                              1                        0   
1                              1                        0   
2                              1                        0   
3                              1                        0   
4                              0                        0   

   native-country_ Vietnam  native-country_ Yugoslavia  
0                        0                           0  
1                        0                           0  
2                        0                           0  
3                        0                           0  
4                        0                           0  

[5 rows x 93 columns]

After removing the catgeorical variables, we see there are now 93 variables (92 predictors and 1 target). 

In the neural network models, the input level will have 92 nodes. In addition, since income has 2 classes, there will be 2 nodes in the output level. The hidden levels and number of nodes in each will vary depending on the model we test.

## Modeling

Now we will try to find the best model that predicts/classifies income. We'll start by creating some neural networks and then compare their performance with some other modeling techniques.

In [ ]:
from numpy import random
random.seed(5)

# define datasets for modeling
X = income_data_clean.drop(['income'],axis=1)
Y = income_data_clean['income']

# split data into test (20%) and train sets (80%)
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, Y, test_size = 0.20, random_state = 5)

### Neural Networks

Let's set up a grid of potential values for some of the neural network parameters. We'll then fit neural network models for all the combinations of these parameter sets.

Some of the parameters we aim to adjust are:
- hidden_layer_sizes: number of hidden layers and nodes in each
- activation: activation function used for hidden layer
- solver: the solver for weight optimization
- alpha: L2 penalty parameter
- learning_rate: learning rate schedule for weight updates (in this lab, "constant" and "adaptive" typically result in similar accuracy rates)


In [ ]:
from sklearn.neural_network import MLPClassifier
random.seed(5)

# define MLP Classifier
mod1 = MLPClassifier(max_iter=100, random_state=5)

# create grid of parameter options
parameter_space = {
    'hidden_layer_sizes': [(100,), (100, 100), (100, 100, 100)],
    'activation': ['relu', 'tanh'],
    'solver': ['adam', 'lbfgs'],
    'alpha': [0.0001, 0.05],
    'learning_rate': ['constant'],
}

In [ ]:
from pandas.core.common import random_state
from sklearn.utils.extmath import randomized_svd
from sklearn.model_selection import GridSearchCV
random.seed(5)

# perform grid search using all the parameters above (with 3 splits)
clf = GridSearchCV(mod1, parameter_space, n_jobs = -1, cv = 3)
clf.fit(train_X, train_y)

Now that we've fit all the models to the train set, we can get the results to determine which set of parameters performed the best. We'll look at the accuracy for each set of parameters.

In [ ]:
random.seed(5)
# All results
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))

0.609 (+/-0.520) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100,), 'learning_rate': 'constant', 'solver': 'adam'}
0.636 (+/-0.004) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100,), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.609 (+/-0.520) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.628 (+/-0.017) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.776 (+/-0.024) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.758 (+/-0.000) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.605 (+/-0.513) for {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100,), 'learning_rate': 'constant', 'solver': 'adam'}
0.636 (

In [ ]:
# get best performing results 
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
  if mean > 0.77:
    print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))

0.776 (+/-0.024) for {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.780 (+/-0.025) for {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.799 (+/-0.002) for {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.792 (+/-0.008) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (100,), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.791 (+/-0.011) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.771 (+/-0.009) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
0.792 (+/-0.003) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 

In [ ]:
random.seed(5)
# Show best parameter set
print('Best parameters found:\n', clf.best_params_)


Best parameters found:
 {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100, 100, 100), 'learning_rate': 'constant', 'solver': 'adam'}


We can see that the best Neural Network model uses:
- Activation Function: relu (rectified linear unit function)
- alpha = 0.05
- 3 hidden layers with 100 nodes in each
- contant learning rate
- adam solver (stochastic gradient-based optimizer)

It has an **accuracy of 0.799**.

Now that we've found the best performing neural network, let's use the model to predict on the test set.

In [ ]:
from sklearn.metrics import accuracy_score
random.seed(5)

# make predictions
pred_y = clf.predict(test_X)

NN_accuracy = accuracy_score(test_y, pred_y)
print('Accuracy: %.2f%%' % (NN_accuracy * 100))

from sklearn.metrics import classification_report
print('Results on the test set:')
print(classification_report(test_y, pred_y))

Accuracy: 79.75%
Results on the test set:
                  precision    recall  f1-score   support

greater than 50K       0.85      0.18      0.30      1548
   less than 50K       0.79      0.99      0.88      4965

        accuracy                           0.80      6513
       macro avg       0.82      0.58      0.59      6513
    weighted avg       0.81      0.80      0.74      6513



From the best model with the parameter set discussed above, we see an overall **80% accuracy**. This is somewhat high, so we could use it to predict whether an individual's income is above $50,000. However, we can also explore other types of models that may predict with higher accuracy.

### Boosting

Let's compare the performance of the neural networks above to some boosting models.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
random.seed(5)

# AdaBoost Classifier
AdaBoost = AdaBoostClassifier(n_estimators=500, random_state=5)
# fit model to train data
AdaBoost.fit(train_X, train_y)

# predict model on test data
Ada_preds = AdaBoost.predict(test_X)

# evaluate model
Ada_accuracy = accuracy_score(test_y, Ada_preds)
print('The accuracy is: ',Ada_accuracy*100,'%')

The accuracy is:  87.07200982650085 %


We can see that the AdaBoost classifier predicts with **87% accuracy**. This is better than the best performing neural networks model.

In [ ]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score
random.seed(5)

# define XGBoost classifier
XGB_model = XGBClassifier(learning_rate=0.5, min_child_weight=6, random_state=5)
XGB_model.fit(train_X, train_y)

# predict model on test data
XGB_preds = XGB_model.predict(test_X)

# evaluate model
XGB_accuracy = accuracy_score(test_y, XGB_preds)
print('Accuracy: %.2f%%' % (XGB_accuracy * 100))

Accuracy: 87.41%


The XGBoost classifier performs about the same as the AdaBoost classifier, with only slightly higher accuracy (**87.4%**). Both Boosting Models predict with much higher accuracy than the Neural Network Models. It would probably be better to use one of these boosting models for predicting income.

Note that boosting models are iterative, which means they continually try to improve themselves until they get the best possible model. Because of this, it is likely the boosting models would out-perform the neural networks.

### Bagging

Let's do the same as we did above with the boosting models, but now using a bagging ensemble model to predict income.

In [22]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
random.seed(5)

# Bagging Classifier
Bag = BaggingClassifier(n_estimators = 100, random_state=5)
# fit model to train data
Bag.fit(train_X, train_y)

# predict model on test data
Bag_preds = Bag.predict(test_X)

# evaluate model
Bag_accuracy = accuracy_score(test_y, Bag_preds)
print('The accuracy is: ',Bag_accuracy*100,'%')

The accuracy is:  85.1527713803163 %


The Bagging Classifier also performs better than the best Neural Network model, with an **accuracy of 85%**. However, it still does not perform as well as the boosting models.

## Best Model

We found the best **Neural Network** model, which predicted whether an individual has over 50K with **80% accuracy**. This is fairly high, but we also found that other modeling techniques result in a higher accuracy. The **XGBoost model** predicted income with **87.4% accuracy**, which is much better than the neural network.

We'll first use the best neural network model to make income predictions for the entire dataset. Then we will create predictions for income using the best performing model overall: XGBoost model.

In [23]:
random.seed(5)
# make predictions from Neural Network
predicted_income_NN = clf.predict(X)
predicted_income_NN = pd.DataFrame(predicted_income_NN, columns = ['predicted_income_NN'])
final_Y = predicted_income_NN.merge(Y, left_index = True, right_index = True)

# make predictions from XGBoost
predicted_income_XGB = XGB_model.predict(X)
predicted_income_XGB = pd.DataFrame(predicted_income_XGB, columns = ['predicted_income_XGB'])
final_Y = predicted_income_NN.merge(final_Y, left_index = True, right_index = True)

final_Y.head()

predicted_income_NN_x predicted_income_NN_y         income
0         less than 50K         less than 50K  less than 50K
1         less than 50K         less than 50K  less than 50K
2         less than 50K         less than 50K  less than 50K
3         less than 50K         less than 50K  less than 50K
4         less than 50K         less than 50K  less than 50K

In [24]:
from sklearn.metrics import accuracy_score
accuracy_NN = accuracy_score(Y, predicted_income_NN)
print('NN accuracy: ', accuracy_NN)

accuracy_XGB = accuracy_score(Y, predicted_income_XGB)
print('XGB accuracy: ', accuracy_XGB)

NN accuracy:  0.7989619483431099
XGB accuracy:  0.8810540216823808


We see that on the entire dataset, the best Neural Network model classifies income with 79.9% accuracy, and the XGBoost model classifies income with 88.1% accuracy. Both do a pretty good job, but the XGBoost is ultimately the best model to use for predicting whether an individual makes over $50,000.